### PR TITLE
Hapi Refactor

### PR DESCRIPTION
**Pull request by kenkouot**
While working on a separate feature, I noticed that our routes file has really grown to be untenable. This is my fault as the patterns that I set when first working on the file didn't scale too well in readability or separation of concerns. This PR will refactor `routes.ts` to be declarative and will move all the business logic to `facets/` and `lib/` for common functions. This also moves the route bounding to `app.ts` and not individually in `routes.ts`. Overall, I think this is a great improvement.

Note, this PR changes have been intended for awhile but after seeing the [npm website](https://github.com/npm/newww) refactor to Hapi, I decided to adopt their patterns to clean up our code.

### CODE REVIEW

**Comment by kenkouot** [comment]
This is ready for you guys to start CR'ing @mklucsarits @hakubo

File: server/facets/showArticle.ts [hunk 1]
```diff
@@ -0,0 +1,122 @@
+/// <reference path="../../typings/hapi/hapi.d.ts" />
+
+import Article = require('../lib/Article');
+import Utils = require('../lib/Utils');
+import Tracking = require('../lib/Tracking');
+import Caching = require('../lib/Caching');
+import localSettings = require('../../config/localSettings');
+
+export = showArticle;
```
Comment (mklucsarits) [inline server/facets/showArticle.ts]: I think the export should be at the end of the file, to keep with convention.

File: server/facets/api/presenters/resultWrapper.ts [hunk 2]
```diff
@@ -0,0 +1,7 @@
+function resultWrapper (err: any, result: any) {
```
Comment (mklucsarits) [inline server/facets/api/presenters/resultWrapper.ts]: Why not just name this function wrapResult, if that's how it's being called?

File: server/facets/showArticle.ts [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+
+var cachingTimes = {
+	enabled: false,
+	cachingPolicy: Caching.Policy.Private,
+	varnishTTL: Caching.Interval.standard,
+	browserTTL: Caching.Interval.default
+};
+
+function showArticle (request: Hapi.Request, reply: Hapi.Response) {
```
Comment (mklucsarits) [inline server/facets/showArticle.ts]: Please add :void to the end of this declaration, since it seems like that's what we're doing now.

**Comment by mklucsarits** [comment]
Please add :void as the return type for the functions that need it in these files, since you're already updating all of them.

File: server/routes.ts [hunk 4]
```diff
@@ -1,340 +1,82 @@
...
+		handler: require('./facets/api/search').get
+	}
+];
 
-			reply('Server status is: OK')
-				.header('X-Memory', String(memoryUsage.rss))
-				.header('X-Uptime', String(~~process.uptime()))
-				.code(200);
-		}
+indexRoutePaths = [
```
Comment (mklucsarits) [inline server/routes.ts]: I don't think "indexRoutes" is an appropriate name anymore for the article paths. We should start to make the app more non-article-centric where possible. Can it be renamed to articleRoutes maybe?

File: server/lib/Utils.ts [hunk 5]
```diff
@@ -1,169 +1,183 @@
...
+export function getCachedWikiDomainName (localSettings: LocalSettings, host: string): string {
+	var wikiDomain: string;
 
-	/**
-	 * @desc Removes the port from hostname
-	 *
-	 * @param {string} host
-	 * @returns {string}
-	 */
-	export function clearHost (host: string): string {
```
Comment (mklucsarits) [inline server/lib/Utils.ts]: This is a weird function name for what it does.

Comment (kenkouot) [inline server/lib/Utils.ts]: I'll fix this in another PR

**Comment by mklucsarits** [comment]
LGTM

**Pull request merged**